---
title: "MVMR_loneliness"
date: "31/10/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

# install and load relevant packages
# install.packages("remotes")
# remotes::install_github("MRCIEU/TwoSampleMR")
library(TwoSampleMR)
library(tidyverse)
library("readr")
library(ggplot2)

# define directories
input = "/path/to/input/data/"
output = "/path/to/output/"
```


### UNIVARIABLE MR: CM -> SHZ

```{r}
# read in Childhood maltreatment Warrier et al. (2021) SNPs 
cm_original <- readxl::read_excel(paste0(input, "ind_snps_maltreatment_original14.xlsx"))

# rename the columns to match what Two Sample MR package is expecting 
cm_original <- cm_original %>% rename(position = 'pos', beta = 'Beta', se = 'SE')
head(cm_original)


# format exposure data
cm_exp_dat <- format_data(cm_original, type="exposure")
cm_exp_dat$exposure <- "maltreatment" # define exposure as maltreatment
head(cm_exp_dat)
```

```{r}
# obtain outcome data using local GWAS summary data
outcome_dat <- read_outcome_data(
  snps = cm_exp_dat$SNP,
  filename = paste0(input, "PGC3_SCZ_wave3.european.autosome.public.v3.vcf.tsv.gz"),
  sep = "\t",
  snp_col = "ID",
  beta_col = "BETA",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "MAF", # originally FCON
  pval_col = "PVAL"
)
outcome_dat$outcome <- 'schizophrenia' 

# which SNPs not in outcome data
ind <- which(!(cm_exp_dat$SNP %in% outcome_dat$SNP))
cm_exp_dat$SNP[ind]
# "rs61818983" "rs77987546" "rs5928362"  "rs6633421" 
cm_exp_dat[ind,]

################################
# HARMONISE DATA
################################

dat <- harmonise_data(
  exposure_dat = cm_exp_dat, 
  outcome_dat = outcome_dat
)
# Harmonising maltreatment (c823Dy) and schizophrenia (gRth8Y)
# Removing the following SNPs for being palindromic with intermediate allele frequencies:
# rs4305836

# calculate F statistic 
# pre-calculate r.outcome using get_r_from_lor() for binary traits
dat$r.outcome <- get_r_from_lor(lor=dat$beta.outcome, af=dat$eaf.outcome, ncase=53386, ncontrol=77258, prevalence=0.01, model='logit') 

# obtain variance in phenotype explained by snps
dat$samplesize.exposure = 185414
dat$samplesize.outcome =  130644 
out <- directionality_test(dat)
out
N = 185414
k = 9 # number of snps 
# for F statistic, use following formula from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8440862/
# F = ((N - K – 1 )  / K) * (R2/(1-R2))
# same formula also in https://academic.oup.com/ije/article/40/3/755/745918 
# Where N represents the sample size, k is the number of SNPs.
Fstat <- dat %>% mutate(F = ((N-k-1)  / k) * (out$snp_r2.exposure /(1-out$snp_r2.exposure ))) 
# 40.38659
#snp_r2.exposure = 0.001956636, snp_r2.outcome = 0.0008836139,correct_causal_direction = TRUE, steiger_pval = 5.770016e-05
write.csv(Fstat, file = paste0(output, 'Fstat-new-uniMR-maltreatment-Shizophrenia-revision.csv'))
write.csv(out, file = paste0(output, 'Steiger-maltreatment-Shizophrenia-revision.csv'))

################################
## PERFORM MR
################################
# for more functions see the tutorial
res <- mr(dat)
# generate odds ratios
res_or <- generate_odds_ratios(res)
write.csv(res_or, file = paste0(output, 'results_MR_CM-Shizophrenia_withOR-revision.csv'))

# Horizontal pleiotropy
### The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. 
mr_pleiotropy_test(dat)

# Heterogeneity statistic
mr_heterogeneity(dat)

# scatter plot
p1 <- my_mr_scatter_plot(res, dat)
p1[[1]]
ggsave(paste0(output,"scatter-maltreatment-scz-revision.png"), dpi = 300, width = 7, height = 6, units = "in")

# forest plot
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
ggsave(paste0(output, "singlesnp-maltreatment-scz-revision.png"))

# Leave-one-out plot
res_loo <- mr_leaveoneout(dat)
p3 <- my_mr_leaveoneout_plot(res_loo)
p3[[1]]
ggsave(paste0(output, "leaveoneout-maltreatment-scz-revision.png"), dpi = 300, width = 7, height = 6, units = "in")

# Funnel plot
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
ggsave(paste0(output, "funnel-plot-maltreatment-scz-revision.png"))

```


### UNIVARIABLE MR: loneliness -> SCZ

```{r}
# read in loneliness SNPs 
loneliness <- readxl::read_excel(paste0(input, "ind_snps_loneliness_MTAG.xlsx"))

# format exposure data
loneliness_exp_dat <- format_data(loneliness, type="exposure")
loneliness_exp_dat$exposure <- "loneliness" # define exposure as maltreatment
head(loneliness_exp_dat)


################################
# OUTCOME DATA
################################

# Using local GWAS summary data

outcome_dat <- read_outcome_data(
  snps = loneliness_exp_dat$SNP,
  filename = paste0(input, "PGC3_SCZ_wave3.european.autosome.public.v3.vcf.tsv.gz"),
  sep = "\t",
  snp_col = "ID",
  beta_col = "BETA",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "MAF", # originally FCON in scz sumstats
  pval_col = "PVAL"
)
outcome_dat$outcome <- 'schizophrenia' 


################################
# HARMONISE DATA
################################

dat <- harmonise_data(
  exposure_dat = loneliness_exp_dat, 
  outcome_dat = outcome_dat
)
nrow(dat) # 15 SNPs remain

# calculate F statistic 
# pre-calculate r.outcome using get_r_from_lor() for binary traits
dat$r.outcome <- get_r_from_lor(lor=dat$beta.outcome, af=dat$eaf.outcome, ncase=53386, ncontrol=77258, prevalence=0.01, model='logit') 

# obtain variance in phenotype explained by snps
dat$samplesize.exposure = 487647
dat$samplesize.outcome =  130644 
out <- directionality_test(dat)
out
#   id.exposure id.outcome   exposure       outcome snp_r2.exposure snp_r2.outcome correct_causal_direction
# 1      li16zn     YDUVNT loneliness schizophrenia     0.001190606   0.0006114892                     TRUE
#   steiger_pval
# 1  0.001683193
nrow(dat %>% filter(mr_keep == TRUE)) # check nr of snps to be kept in MR
N = 487647
k = 15 # number of snps 
# for F statistic, use following formula from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8440862/
# F = ((N - K – 1 )  / K) * (R2/(1-R2))
# same formula also in https://academic.oup.com/ije/article/40/3/755/745918 
# Where N represents the sample size, k is the number of SNPs.
Fstat <- dat %>% mutate(F = ((N-k-1)  / k) * (out$snp_r2.exposure /(1-out$snp_r2.exposure ))) 
# 38.75124
write.csv(Fstat, file = paste0(output, 'Fstat-new-uniMR-loneliness-Shizophrenia-revision.csv'))
write.csv(out, file = paste0(output, 'Steiger-loneliness-Shizophrenia-revision.csv'))

################################
## PERFORM MR
################################
# for more functions see the tutorial
res <- mr(dat)
res_or <- generate_odds_ratios(res)
write.csv(res_or, file = paste0(output, 'results_MR_loneliness-Shizophrenia_withOR-revision.csv'))

# Horizontal pleiotropy
### The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. 
mr_pleiotropy_test(dat)

# Heterogeneity statistic
mr_heterogeneity(dat)

# scatter plot
p1 <- my_mr_scatter_plot(res, dat)
p1[[1]]
library(ggplot2)
ggsave(paste0(output, "scatter-loneliness-scz-revision.png"), dpi = 300, width = 7, height = 6, units = "in")

# forest plot
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
ggsave(paste0(output, "singlesnp-loneliness-scz-revision.png"))

# Leave-one-out plot
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
ggsave(paste0(output, "leaveoneout-loneliness-scz-revision.png"), dpi = 300, width = 7, height = 6, units = "in")

# Funnel plot
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
ggsave(paste0(output, "funnel-plot-loneliness-scz-revision.png"))

```

### UNIVARIABLE MR: CM -> loneliness (a path)
```{r}
# obtain outcome data using local GWAS summary data
outcome_dat <- read_outcome_data(
  snps = cm_exp_dat$SNP,
  filename = paste0(input, "MTAG_results.txt.gz"),
  sep = "\t",
  snp_col = "snpid",
  beta_col = "mtag_beta",
  se_col = "mtag_se",
  effect_allele_col = "a1",
  other_allele_col = "a2",
  eaf_col = "freq", 
  pval_col = "mtag_pval"
) # CHR BP SNP P A1 A2 BETA SE
outcome_dat$outcome <- 'loneliness' 

# which SNPs not in outcome data
ind <- which(!(cm_exp_dat$SNP %in% outcome_dat$SNP))
cm_exp_dat$SNP[ind]
# "rs61818983" "rs77987546" "rs5928362"  "rs6633421" 
cm_exp_dat[ind,]

################################
# HARMONISE DATA
################################

dat <- harmonise_data(
  exposure_dat = cm_exp_dat, 
  outcome_dat = outcome_dat
)
# Harmonising maltreatment (enTKEy) and loneliness (ulA1F0)
# Removing the following SNPs for being palindromic with intermediate allele frequencies:
# rs4305836

# calculate F statistic 

# obtain variance in phenotype explained by snps
dat$samplesize.exposure = 185414
dat$samplesize.outcome =  487647 
out <- directionality_test(dat)
out
N = 185414
k = 9 # number of snps 
# for F statistic, use following formula from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8440862/
# F = ((N - K – 1 )  / K) * (R2/(1-R2))
# same formula also in https://academic.oup.com/ije/article/40/3/755/745918 
# Where N represents the sample size, k is the number of SNPs.
Fstat <- dat %>% mutate(F = ((N-k-1)  / k) * (out$snp_r2.exposure /(1-out$snp_r2.exposure ))) 
# 40.38659
#snp_r2.exposure = 0.001956636, snp_r2.outcome = 0.0008836139,correct_causal_direction = TRUE, steiger_pval = 5.770016e-05
write.csv(Fstat, file = paste0(output, 'Fstat-new-uniMR-maltreatment-loneliness-revision2-2024.csv'))
write.csv(out, file = paste0(output, 'Steiger-maltreatment-loneliness-revision2-2024.csv'))

################################
## PERFORM MR
################################
# for more functions see the tutorial
res <- mr(dat)
# generate odds ratios
res_or <- generate_odds_ratios(res)
write.csv(res_or, file = paste0(output, 'results_MR_CM-loneliness-revision2-2024.csv'))

# Horizontal pleiotropy
### The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. 
mr_pleiotropy_test(dat)
#  id.exposure id.outcome    outcome     exposure egger_intercept          se     pval
# 1      enTKEy     ulA1F0 loneliness maltreatment      0.01670444 0.009765928 0.130918

# Heterogeneity statistic
mr_heterogeneity(dat)
#   id.exposure id.outcome    outcome     exposure                    method        Q Q_df     Q_pval
# 1      enTKEy     ulA1F0 loneliness maltreatment                  MR Egger 12.44842    7 0.08674352
# 2      enTKEy     ulA1F0 loneliness maltreatment Inverse variance weighted 17.65142    8 0.02399722

# scatter plot
source("plot_functions.R")
p1 <- my_mr_scatter_plot(res, dat)
p1[[1]]
ggsave(paste0(output,"scatter-maltreatment-loneliness-revision2-2024.png"), dpi = 300, width = 7, height = 6, units = "in")

# forest plot
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
ggsave(paste0(output, "singlesnp-maltreatment-loneliness-revision2-2024.png"))

# Leave-one-out plot
res_loo <- mr_leaveoneout(dat)
p3 <- my_mr_leaveoneout_plot(res_loo)
p3[[1]]
ggsave(paste0(output, "leaveoneout-maltreatment-loneliness-revision2-2024.png"), dpi = 300, width = 7, height = 6, units = "in")

# Funnel plot
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
ggsave(paste0(output, "funnel-plot-maltreatment-loneliness-revision2-2024.png"))

```


### Preparation for MVMR
```{r eval = FALSE}
# read in independent SNPs for each trait
cm_original <- readxl::read_excel(paste0(input, "ind_snps_maltreatment_original14.xlsx"))
loneliness <- readxl::read_excel(paste0(input, "ind_snps_loneliness_MTAG.xlsx"))

# create a vector of SNPs both phenotypes
snp_set <- c(cm_original$SNP, loneliness$SNP) # 29 SNPs
unique(snp_set)

# extract snp_set from cm data
# read in childhood maltreatment sumstats
cm_raw <- read.delim(paste0(input, 'Retro_prospective_meta_childhoodmaltreatment.txt.gz'), sep = ' ')

library(tidyverse)
cm_data <- cm_raw %>% filter(SNP %in% snp_set) 
cm_data$phenotype <- "Maltreatment"
names(cm_data) <- c("chr", "position", "SNP", "pval", "effect_allele", "other_allele", "beta", "se", "phenotype")
write.csv(cm_data, file = paste0(input, 'cm_data_mvmr_revision.csv'), row.names = F)

# extract snp_set from loneliness data
# read in loneliness sumstats
loneliness_raw <- read.delim(paste0(input, 'MTAG_results.txt.gz'))
head(loneliness_raw)

loneliness_data <- loneliness_raw %>% filter(snpid %in% snp_set) # 25 SNPs here, as still includes 1 palindromic snp (should be 24, 9 for CM + 15 for loneliness)
# rename header
names(loneliness_data) <- c("SNP", "chr", "position", "effect_allele", "other_allele", "not-mtag-z", "N", "eaf", "beta", "se", "z", "pval")
loneliness_data$phenotype <- "Loneliness"
write.csv(loneliness_data, file = paste0(input, 'loneliness_data_mvmr_revision.csv'), row.names = F)
```

### MVMR using TwoSampleMR package 
```{r}
library(TwoSampleMR)
library(tidyverse)
library("readr")

exposure_dat  <- mv_extract_exposures_local(c(paste0(input, 'loneliness_data_mvmr_revision.csv'),
                                              paste0(input, 'cm_data_mvmr_revision.csv')),
                                            phenotype_col = 'phenotype',
                                            harmonise_strictness=2,sep = ',') # 46 SNPs

outcome_dat <- read_outcome_data(
  snps = exposure_dat$SNP,
  filename = paste0(input, "PGC3_SCZ_wave3.european.autosome.public.v3.vcf.tsv.gz"),
  sep = "\t",
  snp_col = "ID",
  beta_col = "BETA",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "MAF",
  pval_col = "PVAL"
)
outcome_dat$outcome <- 'schizophrenia' 
# 23 SNPs

# Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat) 
res_mvmr <- mv_multiple(mvdat) # fits all exposures together (recommended as default in TwoSampleMR)

# generate OR
res_mvmr <- generate_odds_ratios(res_mvmr$result)
write.csv(res_mvmr, file = paste0(output, 'results_MVMR_loneliness-maltreatment-schizophrenia-revision.csv'))
```


### MVMR Egger
```{r}
# from https://rdrr.io/cran/MendelianRandomization/man/mr_mvegger.html
# from https://rdrr.io/cran/MendelianRandomization/f/inst/doc/Vignette_MR.pdf 

library(MendelianRandomization)
MRMVInputObject <- mr_mvinput(bx = mvdat$exposure_beta,
                              bxse = mvdat$exposure_se,
                              by = mvdat$outcome_beta,
                              byse = mvdat$outcome_se)

MRMVInputObject 

MRMVObject <- mr_mvivw(MRMVInputObject, model = "default",
                          correl = FALSE,
                          distribution = "normal",
                          alpha = 0.05)

MRMVObject # exactly same output as from TwoSampleMR package
# MRMVObject$Pvalue 

# run mvmr mr-egger
MVEgger <- mr_mvegger(MRMVInputObject, orientate = 2)

MVEgger_df <- as.data.frame(cbind(exposure = MVEgger$Exposure, b = MVEgger$Estimate, se = MVEgger$StdError.Est, pval = MVEgger$Pvalue.Est))

MVEgger_df$b <- as.numeric(MVEgger_df$b)
MVEgger_df$se <- as.numeric(MVEgger_df$se)
MVEgger_df <- generate_odds_ratios(MVEgger_df)
write.csv(MVEgger_df, file = paste0(output, 'results_MVEgger_maltreatment-loneliness-schizophrenia-revision.csv'))
######################################
```


### Use Eleonora Sanders MVMR package 
Perform 4 recommended steps for MVMR
from: https://onlinelibrary.wiley.com/doi/full/10.1002/sim.9133
```{r}
#install.packages("remotes")
library(remotes)
#install_github("WSpiller/MVMR", build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE)

#To update the package just run the remotes::install_github("WSpiller/MVMR", build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE) command again.
vignette("MVMR") # vigentte: https://github.com/WSpiller/MVMR/blob/master/vignettes/MVMR.rmd
# package obtained from: https://github.com/WSpiller/MVMR/
library(MVMR)

my.F  <- format_mvmr(BXGs = mvdat$exposure_beta,
                      BYG = mvdat$outcome_beta,
                      seBXGs = mvdat$exposure_se,
                      seBYG = mvdat$outcome_se,
                      RSID = rownames(mvdat$exposure_beta))
                            
sres <- strength_mvmr(r_input = my.F, gencov = 0) # Fixing covariance at 0
# Conditional F-statistics for instrument strength
# 
#             exposure1 exposure2
# F-statistic  10.15759   12.2672 

# setting `mvmrcov` to zero should be appropriate given the minimum sample overlap with outcome.
```


#### Test for horizontal pleiotropy using conventional Q-statistic estimation

```{r}
# heterogeneity is indicative of a violation of the exclusion restriction assumption 
pres <- pleiotropy_mvmr(r_input = my.F, gencov = 0)
```

#### Product of coefficients 
```{r}
# calculate indirect effect 
apath = 0.225111886411753
bpath = 0.436560049716838
a.se = 0.0489899991170715
b.se = 0.477069983297695
indirect.effect = apath * bpath # 0.09827486

# calculate SE of indirect effect 
se.indirect.effect = sqrt(apath^2 * b.se^2 + bpath^2 * a.se^2)
# 0.109503

# Calculate the z-score
z <- indirect.effect / se.indirect.effect 
# 0.8974628

# Calculate the two-tailed p-value
p_value <- 2 * (1 - pnorm(abs(z))) # same as 2*pnorm(-abs(z))
# 0.369472

# MEDIATION (indirect effect):
# beta = 0.10 (se = 0.11), p = 0.369
```