---
title: "MVMR_loneliness"
date: "4/18/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

# install and load relevant packages
# install.packages("remotes")
# remotes::install_github("MRCIEU/TwoSampleMR")
library(TwoSampleMR)
library(tidyverse)
library("readr")

# define directories
input = "path/to/input/data"
output = "path/to/output/"
```


### UNIVARIABLE MR: CM -> SHZ

```{r}
# read in Childhood maltreatment Warrier et al. (2021) SNPs 
cm_original <- readxl::read_excel(paste0(input, "ind_snps_maltreatment_original14.xlsx"))
#cm <- read.csv('proxy_cm_loci_14.csv')

# rename the columns to match what Two Sample MR package is expecting 
cm_original <- cm_original %>% rename(position = 'pos', beta = 'Beta', se = 'SE')
head(cm_original)


# format exposure data
cm_exp_dat <- format_data(cm_original, type="exposure")
cm_exp_dat$exposure <- "maltreatment" # define exposure as maltreatment
head(cm_exp_dat)
```


```{r eval = FALSE}
################################
# OUTCOME DATA
################################
# read in schizophrenia data 
schizo <- read.delim("paste0(input, schizophrenia_rsid.txt.gz"))
#to calculate the natural log, use the log() function (does the opposite of exponentiate) 
schizo <- schizo %>% mutate(log_beta = log(OR))
# save the new file 
# save cleaned data
write.table(schizo, file = paste0(input, "schizophrenia_rsid_logbeta.txt"), row.names = FALSE, quote = FALSE, sep = '\t')

# convert OR to beta
# calculate beta coefficient and its error from OR from GWAS summary Statisitcs
# Beta= log(OR)
# SE of Beta
# You can calculate the upper and lower bounds for the OR by
# Upper bound = OR + se(OR) x 1.96
# Lower bound = OR - se(OR) x 1.96
# Log(upper bound of OR) = upper bound of beta
# Log(lower bound of OR) = lower bound of beta
# If you have the 95% C.I of beta, then calculating the SE(beta) is quite simple!
# for example if beta = 0.5 and the upper C.I is 0.6 then
# upper C.I of beta = beta + se(beta) x 1.96
# 0.6 = 0.5 + se x 1.96
# Assuming the confidence interval being used is 95%.
```

```{r}
# Using local GWAS summary data
outcome_dat <- read_outcome_data(
  snps = cm_exp_dat$SNP,
  filename = paste0(input, "schizophrenia_rsid_logbeta.txt.gz"),
  sep = "\t",
  #sep = " ",
  snp_col = "SNP",
  beta_col = "log_beta",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "Freq.A1",
  pval_col = "P"
)
outcome_dat$outcome <- 'schizophrenia' 


################################
# HARMONISE DATA
################################

dat <- harmonise_data(
  exposure_dat = cm_exp_dat, 
  outcome_dat = outcome_dat
)
# Harmonising maltreatment (AwUzQA) and schizophrenia (fAavKh)
# Removing the following SNPs for being palindromic with intermediate allele frequencies:
#   rs4305836, rs5928362

nrow(dat) # 12 SNPs remain

# calculate F statistic (From https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8440862/)
# eaf.exposure eaf.outcome 
#R2 = 2*beta * MAF * (1-MAF) 
# where beta represents the effect value of the genetic variant in the exposure and MAF represents the effect allele frequency
#F = ((N - K â€“ 1 )  / K) * (R2/(1-R2))
# Where N represents the sample size, k is the number of SNPs.

Fstat <- dat %>% mutate(R2 = abs(2*beta.exposure * eaf.exposure * (1-eaf.exposure)))
Fstat <- Fstat %>% mutate(F = ((185414-12-1)  / 12) * (R2/(1-R2)))

#write.csv(Fstat, file = paste0(output, 'Fstat-uniMR-maltreatment-Shizophrenia.csv'))

################################
## PERFORM MR
################################
# for more functions see the tutorial
res <- mr(dat)
# generate odds ratios
res_or <- generate_odds_ratios(res)
#write.csv(res_or, file = paste0(output, 'results_MR_CM-Shizophrenia_withOR.csv'))

# Horizontal pleiotropy
### The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. 
mr_pleiotropy_test(dat)

# Heterogeneity statistic
mr_heterogeneity(dat)

# scatter plot
p1 <- mr_scatter_plot(res, dat)
p1[[1]]

library(ggplot2)
ggsave(paste0(output,"scatter-maltreatment-scz.pdf"))

# forest plot
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
ggsave(paste0(output, "singlesnp-maltreatment-scz.pdf"))

# Leave-one-out plot
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
ggsave(paste0(output, "leaveoneout-maltreatment-scz.pdf"))

# Funnel plot
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
ggsave(paste0(output, "funnel-plot-scz.pdf"))

```


### UNIVARIABLE MR: loneliness -> SHZ

```{r}
# read in loneliness SNPs 
loneliness <- readxl::read_excel(paste0(input, "ind_snps_loneliness_MTAG.xlsx"))

# format exposure data
loneliness_exp_dat <- format_data(loneliness, type="exposure")
loneliness_exp_dat$exposure <- "loneliness" # define exposure as maltreatment
head(loneliness_exp_dat)


################################
# OUTCOME DATA
################################

# Using local GWAS summary data

outcome_dat <- read_outcome_data(
  snps = loneliness_exp_dat$SNP,
  filename = paste0(input,  "schizophrenia_rsid_logbeta.txt.gz"),
  sep = "\t",
  #sep = " ",
  snp_col = "SNP",
  beta_col = "log_beta",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "Freq.A1",
  pval_col = "P"
)
outcome_dat$outcome <- 'schizophrenia' 


################################
# HARMONISE DATA
################################

dat <- harmonise_data(
  exposure_dat = loneliness_exp_dat, 
  outcome_dat = outcome_dat
)
nrow(dat) # 15 SNPs remain

# calculate F statistic 
Fstat <- dat %>% mutate(R2 = abs(2*beta.exposure * eaf.exposure * (1-eaf.exposure)))
Fstat <- Fstat %>% mutate(F = ((487647-15-1)  / 15) * (R2/(1-R2)))

Fstat %>% mutate(F = 32508.73 * (R2/(1-R2)))

write.csv(Fstat, file = paste0(output, 'Fstat-uniMR-loneliness-Shizophrenia.csv'))

################################
## PERFORM MR
################################
# for more functions see the tutorial
res <- mr(dat)
res_or <- generate_odds_ratios(res)
#write.csv(res_or, file = paste0(output, 'results_MR_loneliness-Shizophrenia_withOR.csv'))

# Horizontal pleiotropy
### The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. 
mr_pleiotropy_test(dat)

# Heterogeneity statistic
mr_heterogeneity(dat)

# scatter plot
p1 <- mr_scatter_plot(res, dat)
p1[[1]]
library(ggplot2)
ggsave(paste0(output, "scatter-loneliness-scz.pdf"))

# forest plot
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
ggsave(paste0(output, "singlesnp-loneliness-scz.pdf"))

# Leave-one-out plot
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
ggsave(paste0(output, "leaveoneout-loneliness-scz.pdf"))

# Funnel plot
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
ggsave(paste0(output, "funnel-plot-scz.pdf"))

```

### Preparation for MVMR
```{r eval = FALSE}
# read in independent SNPs for each trait
cm_original <- readxl::read_excel(paste0(input, "ind_snps_maltreatment_original14.xlsx"))
loneliness <- readxl::read_excel(paste0(input, "ind_snps_loneliness_MTAG.xlsx"))

# create a vector of SNPs both phenotypes
snp_set <- c(cm_original$SNP, loneliness$SNP) # 29 SNPs
unique(snp_set)

# extract snp_set from cm data
# read in childhood maltreatment sumstats
cm_raw <- read.delim(paste0(input, 'Retro_prospective_meta_childhoodmaltreatment.txt.gz'), sep = ' ')

library(tidyverse)
cm_data <- cm_raw %>% filter(SNP %in% snp_set) 
cm_data$phenotype <- "Maltreatment"
names(cm_data) <- c("chr", "position", "SNP", "pval", "effect_allele", "other_allele", "beta", "se", "phenotype")
#write.csv(cm_data, file = paste0(input, 'cm_data_mvmr.csv'), row.names = F)

# extract snp_set from loneliness data
# read in loneliness sumstats
loneliness_raw <- read.delim(paste0(input, 'MTAG_results.txt.gz'))
head(loneliness_raw)

loneliness_data <- loneliness_raw %>% filter(snpid %in% snp_set) # 25 SNPs here only (4 missing in loneliness raw data)
#write.csv(loneliness_data, file = paste0(input, 'loneliness_data_mvmr.csv'), row.names = F)
```

### MVMR using TwoSampleMR package 
```{r}
library(TwoSampleMR)
library(tidyverse)
library("readr")
library(MRPracticals)


exposure_dat  <- mv_extract_exposures_local(c(paste0(input, 'loneliness_data_mvmr.csv'),
                                              paste0(input, 'cm_data_mvmr.csv')),
                                            phenotype_col = 'phenotype',
                                            harmonise_strictness=2,sep = ',') # 46 SNPs

# prepare schizophrenia file so that SNPs are separate (and only keep rows with rsid) 
# gunzip -cd clozuk_pgc2.meta.sumstats.txt.gz | grep -e '^SNP\|^rs' > schizophrenia.txt
# wc -l schizophrenia.txt
# 8167164
#tr -s ': ' '\t' < schizophrenia.txt > schizophrenia2.txt 
#gunzip -cd schizophrenia.txt.gz | tr -s ': ' '\t' > schizophrenia_rsid.txt 
# nano to rename columns and gzip to get the final file
# gzip schizophrenia_rsid.txt  

outcome_dat <- read_outcome_data(
  snps = exposure_dat$SNP,
  filename = paste0(input, "schizophrenia_rsid_logbeta.txt.gz"),
  sep = "\t",
  #sep = " ",
  snp_col = "SNP",
  beta_col = "log_beta",
  se_col = "SE",
  effect_allele_col = "A1",
  other_allele_col = "A2",
  eaf_col = "Freq.A1",
  pval_col = "P"
)
outcome_dat$outcome <- 'schizophrenia' 
# 23 SNPs

# Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat) 
res_mvmr <- mv_multiple(mvdat) # fits all exposures together (recommended as default in TwoSampleMR)

# generate OR
res_mvmr <- generate_odds_ratios(res_mvmr$result)

#write.csv(res_mvmr, file = paste0(output, 'results_MVMR_loneliness-maltreatment-shizophrenia.csv'))
```


### MVMR Egger
```{r}
# from https://rdrr.io/cran/MendelianRandomization/man/mr_mvegger.html
# from https://rdrr.io/cran/MendelianRandomization/f/inst/doc/Vignette_MR.pdf 

library(MendelianRandomization)
MRMVInputObject <- mr_mvinput(bx = mvdat$exposure_beta,
                              bxse = mvdat$exposure_se,
                              by = mvdat$outcome_beta,
                              byse = mvdat$outcome_se)

MRMVInputObject 

MRMVObject <- mr_mvivw(MRMVInputObject, model = "default",
                          correl = FALSE,
                          distribution = "normal",
                          alpha = 0.05)

MRMVObject

MRMVObject <- mr_mvivw(MRMVInputObject)

MRMVObject # exactly same output as from TwoSampleMR package
# MRMVObject$Pvalue 

MVEgger <- mr_mvegger(MRMVInputObject, orientate = 2)

MVEgger_df <- as.data.frame(cbind(exposure = MVEgger$Exposure, b = MVEgger$Estimate, se = MVEgger$StdError.Est, pval = MVEgger$Pvalue.Est))
#write.csv(MVEgger_df, file = paste0(output, 'results_MVEgger_maltreatment-loneliness-shizophrenia.csv'))

MVEgger_df$b <- as.numeric(MVEgger_df$b)
MVEgger_df$se <- as.numeric(MVEgger_df$se)
generate_odds_ratios(MVEgger_df)
######################################
```


### Use Eleonora Sanders MVMR package 
Perform 4 recommended steps for MVMR
from: https://onlinelibrary.wiley.com/doi/full/10.1002/sim.9133
```{r}
#install.packages("remotes")
library(remotes)
#install_github("WSpiller/MVMR", build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE)

#To update the package just run the remotes::install_github("WSpiller/MVMR", build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE) command again.
vignette("MVMR") # vigentte: https://github.com/WSpiller/MVMR/blob/master/vignettes/MVMR.rmd
# package obtained from: https://github.com/WSpiller/MVMR/
library(MVMR)

my.F  <- format_mvmr(BXGs = mvdat$exposure_beta[, c(1,2)],
                      BYG = mvdat$exposure_se[, c(1,2)],
                      seBXGs = mvdat$outcome_beta,
                      seBYG = mvdat$outcome_se,
                      RSID = rownames(mvdat$exposure_beta))
                              
                      
sres <- strength_mvmr(r_input = my.F, gencov = 0)
# Conditional F-statistics for instrument strength
# 
#             exposure1 exposure2
# F-statistic  6.934204  8.751063 (prior run 9.375003  1.012232, maybe diff set)

# setting `mvmrcov` to zero should be appropriate given the minimum sample overlap. 
```


#### Test for horizontal pleiotropy using conventional Q-statistic estimation

```{r}
# heterogeneity is indicative of a violation of the exclusion restriction assumption 
pres <- pleiotropy_mvmr(r_input = my.F, gencov = 0)
```
